File: backend/src/services/auth.service.ts
import bcrypt from 'bcryptjs';
import { prisma } from '../lib/prisma.js';
import { userProvider } from '../providers/userProvider.js';
import { roleProvider } from '../providers/roleProvider.js';
import { refreshTokenProvider } from '../providers/refreshTokenProvider.js';
import { AppError } from '../utils/AppError.js';
import type {
  RegisterBodyType,
  LoginRequest,
  GuestRegisterBodyType,
} from '../schemas/validation.js';
import { AuthResultType, SafeUserType } from '../schemas/auth.js';
import {
  ACCESS_TOKEN_EXPIRY_MS,
  ACCESS_TOKEN_EXPIRY_SECONDS,
  generateAccessToken,
  generateRefreshToken,
  generateTokenFamily,
  hashToken,
  MAX_SESSIONS,
  REFRESH_TOKEN_EXPIRY_SECONDS,
  REFRESH_TOKEN_EXPIRY_DAYS,
} from '../utils/authUtils.js';

function toSafeUser(user: {
  id: bigint;
  email: string | null;
  username: string | null;
  name: string | null;
  img?: string | null;
}): SafeUserType {
  return {
    id: user.id.toString(),
    email: user.email ?? '',
    username: user.username,
    name: user.name,
    img: user.img ?? null,
  };
}

async function issueTokenPair(
  user: { id: bigint; token_version: number; email: string | null },
  roles: string[],
  req: { headers: Record<string, string | string[] | undefined>; ip?: string },
): Promise<{ accessToken: string; refreshToken: string }> {
  const rawRefresh = generateRefreshToken();
  const hashed = hashToken(rawRefresh);
  const family = generateTokenFamily();

  // Enforce max sessions
  const count = await refreshTokenProvider.countActiveSessions(user.id);
  if (count >= MAX_SESSIONS) {
    const oldest = await refreshTokenProvider.findOldestActiveSession(user.id);
    if (oldest)
      await refreshTokenProvider.revokeById(oldest.id, 'session_limit');
  }

  const ip =
    (Array.isArray(req.headers['x-forwarded-for'])
      ? req.headers['x-forwarded-for'][0]
      : req.headers['x-forwarded-for']) ??
    req.ip ??
    null;

  await refreshTokenProvider.create({
    user_id: user.id,
    token: hashed,
    token_family: family,
    device_info: (req.headers['user-agent'] as string) ?? null,
    ip_address: ip as string | null,
    expires_at: new Date(
      Date.now() + Number(REFRESH_TOKEN_EXPIRY_DAYS) * 24 * 60 * 60 * 1000,
    ),
  });

  const accessToken = generateAccessToken(
    user.id,
    user.token_version,
    user.email ?? '',
    roles,
  );
  return { accessToken, refreshToken: rawRefresh };
}

export const authService = {
  async register(
    dto: RegisterBodyType,
    reqCtx: {
      headers: Record<string, string | string[] | undefined>;
      ip?: string;
    } = { headers: {} },
  ): Promise<AuthResultType & { refreshToken: string; refreshTokenExpiresAt: string }> {
    const existing = await userProvider.findByEmail(dto.email);
    if (existing) throw new AppError(409, 'Email already registered');

    const hashed = await bcrypt.hash(dto.password, 12);
    const user = await userProvider.create({
      email: dto.email,
      password: hashed,
      username: dto.username,
      name: dto.name,
    });

    const roleName = dto.role ?? 'CUSTOMER';
    const role = await roleProvider.findRoleByName(roleName);
    if (role) await roleProvider.assignRole(user.id, role.id);

    const roles = [roleName];
    const { accessToken, refreshToken } = await issueTokenPair(
      { id: user.id, token_version: 0, email: user.email },
      roles,
      reqCtx,
    );
    const expiresAt = new Date(Date.now() + ACCESS_TOKEN_EXPIRY_MS);
    const refreshTokenExpiresAt = new Date(
      Date.now() + Number(REFRESH_TOKEN_EXPIRY_DAYS) * 24 * 60 * 60 * 1000,
    );
    return {
      token: accessToken,
      expiresAt: expiresAt.toISOString(),
      refreshToken,
      refreshTokenExpiresAt: refreshTokenExpiresAt.toISOString(),
      user: toSafeUser(user),
      role: roles,
    };
  },

  async guestRegister(dto: GuestRegisterBodyType): Promise<AuthResultType> {
    const existing = await userProvider.findByPhone(dto.phone);
    const user =
      existing ??
      (await userProvider.createGuest({
        name: dto.name,
        phone: dto.phone,
        is_guest: true,
      }));
    const roles = ['GUEST'];
    const { accessToken } = await issueTokenPair(
      { id: user.id, token_version: 0, email: user.email },
      roles,
      { headers: {}, ip: undefined },
    );
    const expiresAt = new Date(Date.now() + ACCESS_TOKEN_EXPIRY_MS);
    return {
      token: accessToken,
      expiresAt: expiresAt.toISOString(),
      user: toSafeUser(user),
      role: roles,
    };
  },

  async login(
    dto: LoginRequest,
    reqCtx: {
      headers: Record<string, string | string[] | undefined>;
      ip?: string;
    },
  ): Promise<
    AuthResultType & { refreshToken: string; refreshTokenExpiresAt: string }
  > {
    const user = await userProvider.findByEmail(dto.email);
    if (!user) throw new AppError(401, 'Invalid email or password');

    const valid = await bcrypt.compare(dto.password, user.password!);
    if (!valid) throw new AppError(401, 'Invalid email or password');

    const roles = (user.roles as any[]).map((ur) => ur.role.name);
    const primaryRoles = roles.length > 0 ? roles : ['CUSTOMER'];

    const fullUser = await (prisma.user.findUnique as any)({
      where: { id: user.id },
      select: { token_version: true },
    });
    const tokenVersion = (fullUser as any)?.token_version ?? 0;

    const { accessToken, refreshToken } = await issueTokenPair(
      { id: user.id, token_version: tokenVersion, email: user.email },
      primaryRoles,
      reqCtx,
    );
    const expiresAt = new Date(Date.now() + ACCESS_TOKEN_EXPIRY_MS);
    const refreshTokenExpiresAt = new Date(
      Date.now() + Number(REFRESH_TOKEN_EXPIRY_DAYS) * 24 * 60 * 60 * 1000,
    );
    return {
      token: accessToken,
      expiresAt: expiresAt.toISOString(),
      refreshToken,
      refreshTokenExpiresAt: refreshTokenExpiresAt.toISOString(),
      user: toSafeUser(user),
      role: primaryRoles,
    };
  },

  async refresh(
    rawToken: string,
    reqCtx: {
      headers: Record<string, string | string[] | undefined>;
      ip?: string;
    },
  ): Promise<{
    accessToken: string;
    expiresAt: string;
    expiresIn: number;
    refreshToken: string;
    refreshExpires: number;
  }> {
    const hashed = hashToken(rawToken);
    const stored = await refreshTokenProvider.findByToken(hashed);
    if (!stored) throw new AppError(401, 'Invalid refresh token');

    if (stored.revoked) {
      await refreshTokenProvider.revokeFamily(stored.token_family);
      throw new AppError(401, 'Security alert: token reuse detected');
    }

    if (stored.expires_at < new Date()) {
      throw new AppError(401, 'Refresh token expired');
    }

    const user = await prisma.user.findUnique({
      where: { id: stored.user_id },
      include: { roles: { include: { role: true } } },
    });
    if (!user || user.deleted || !user.active)
      throw new AppError(401, 'User not found');

    await refreshTokenProvider.revokeById(stored.id, 'rotation');

    const roles = (user.roles as any[]).map((ur) => ur.role.name);
    const rawNew = generateRefreshToken();
    const hashedNew = hashToken(rawNew);
    const ip =
      (Array.isArray(reqCtx.headers['x-forwarded-for'])
        ? reqCtx.headers['x-forwarded-for'][0]
        : reqCtx.headers['x-forwarded-for']) ??
      reqCtx.ip ??
      null;

    await refreshTokenProvider.create({
      user_id: user.id,
      token: hashedNew,
      token_family: stored.token_family,
      device_info: (reqCtx.headers['user-agent'] as string) ?? null,
      ip_address: ip as string | null,
      expires_at: new Date(
        Date.now() + Number(REFRESH_TOKEN_EXPIRY_DAYS) * 24 * 60 * 60 * 1000,
      ),
    });

    const accessToken = generateAccessToken(
      user.id,
      (user as any).token_version ?? 0,
      user.email ?? '',
      roles,
    );
    const expiresAt = new Date(Date.now() + ACCESS_TOKEN_EXPIRY_MS);

    return {
      accessToken,
      expiresAt: expiresAt.toISOString(),
      expiresIn: ACCESS_TOKEN_EXPIRY_SECONDS,
      refreshToken: rawNew,
      refreshExpires: REFRESH_TOKEN_EXPIRY_SECONDS,
    };
  },

  async logout(rawToken: string): Promise<void> {
    if (!rawToken) return;
    const hashed = hashToken(rawToken);
    const stored = await refreshTokenProvider.findByToken(hashed);
    if (!stored) return;
    await refreshTokenProvider.revokeById(stored.id, 'logout');
  },

  async logoutAll(userId: bigint): Promise<void> {
    await (prisma as any).user.update({
      where: { id: userId },
      data: { token_version: { increment: 1 } },
    });
    await refreshTokenProvider.revokeAllForUser(userId);
  },

  async changePassword(
    userId: bigint,
    currentPassword: string,
    newPassword: string,
    reqCtx: {
      headers: Record<string, string | string[] | undefined>;
      ip?: string;
    },
  ): Promise<{ accessToken: string; refreshToken: string }> {
    const user = await prisma.user.findUnique({
      where: { id: userId },
      include: { roles: { include: { role: true } } },
    });
    if (!user) throw new AppError(404, 'User not found');

    const valid = await bcrypt.compare(currentPassword, user.password!);
    if (!valid) throw new AppError(401, 'Current password is incorrect');

    if (newPassword.length < 8)
      throw new AppError(400, 'New password must be at least 8 characters');

    const newHash = await bcrypt.hash(newPassword, 12);
    await prisma.user.update({
      where: { id: userId },
      data: { password: newHash },
    });

    // Logout all — increment version + revoke all tokens
    await (prisma as any).user.update({
      where: { id: userId },
      data: { token_version: { increment: 1 } },
    });
    await refreshTokenProvider.revokeAllForUser(userId);

    // Issue fresh tokens for current session
    const updatedUser = await prisma.user.findUnique({ where: { id: userId } });
    const roles = (user.roles as any[]).map((ur) => ur.role.name);
    const { accessToken, refreshToken } = await issueTokenPair(
      {
        id: userId,
        token_version: (updatedUser as any)!.token_version ?? 0,
        email: user.email,
      },
      roles,
      reqCtx,
    );
    return { accessToken, refreshToken };
  },

  async me(userId: string): Promise<SafeUserType & { role: string[] }> {
    const user = await userProvider.findById(Number(userId));
    if (!user) throw new AppError(404, 'User not found');
    const roles = (user.roles as any[]).map((ur: any) => ur.role.name);
    return {
      ...toSafeUser(user),
      role: roles.length > 0 ? roles : ['CUSTOMER'],
    };
  },
};

File: backend/app/services/auth_service.py
from datetime import datetime, timedelta, timezone
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update
from app.providers.user_provider import user_provider
from app.providers.role_provider import role_provider
from app.providers.refresh_token_provider import refresh_token_provider
from app.utils.app_error import AppError
from app.utils.auth_utils import (
    hash_token, generate_refresh_token, generate_token_family,
    generate_access_token, ACCESS_TOKEN_EXPIRY_MINUTES, ACCESS_TOKEN_EXPIRY_SECONDS,
    REFRESH_TOKEN_EXPIRY_DAYS, REFRESH_TOKEN_EXPIRY_SECONDS, MAX_SESSIONS,
    hash_password, verify_password,
)
from typing import Optional

def to_safe_user(user) -> dict:
    return {
        "id": str(user.id),
        "email": user.email or "",
        "username": user.username,
        "name": user.name,
        "img": user.img,
    }

async def _issue_token_pair(
    db: AsyncSession,
    user,
    roles: list[str],
    req_ctx: dict,
) -> dict:
    raw_refresh = generate_refresh_token()
    hashed = hash_token(raw_refresh)
    family = generate_token_family()

    # Enforce max sessions
    count = await refresh_token_provider.count_active_sessions(db, user.id)
    if count >= MAX_SESSIONS:
        oldest = await refresh_token_provider.find_oldest_active_session(db, user.id)
        if oldest:
            await refresh_token_provider.revoke_by_id(db, oldest.id, "session_limit")

    # Extract IP
    headers = req_ctx.get("headers", {})
    ip = headers.get("x-forwarded-for") or req_ctx.get("ip")
    if isinstance(ip, list):
        ip = ip[0] if ip else None
    device_info = headers.get("user-agent")

    expires_at = datetime.now(timezone.utc) + timedelta(days=REFRESH_TOKEN_EXPIRY_DAYS)
    await refresh_token_provider.create(
        db,
        user_id=user.id,
        token=hashed,
        token_family=family,
        device_info=device_info,
        ip_address=ip,
        expires_at=expires_at,
    )

    access_token = generate_access_token(user.id, user.token_version, user.email or "", roles)
    return {"accessToken": access_token, "refreshToken": raw_refresh}

async def register(db: AsyncSession, dto: dict, req_ctx: dict = {}) -> dict:
    existing = await user_provider.find_by_email(db, dto["email"])
    if existing:
        raise AppError(409, "Email already registered")
    hashed_pw = hash_password(dto["password"])
    user = await user_provider.create(db, email=dto["email"], password=hashed_pw, username=dto.get("username"), name=dto.get("name"))
    role_name = dto.get("role") or "CUSTOMER"
    role = await role_provider.find_by_name(db, role_name)
    if role:
        await role_provider.assign_role(db, user.id, role.id)
    roles = [role_name]
    tokens = await _issue_token_pair(db, user, roles, req_ctx)
    expires_at = (datetime.now(timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRY_MINUTES)).isoformat()
    refresh_expires_at = (datetime.now(timezone.utc) + timedelta(days=REFRESH_TOKEN_EXPIRY_DAYS)).isoformat()
    return {"token": tokens["accessToken"], "expiresAt": expires_at, "refreshToken": tokens["refreshToken"], "refreshTokenExpiresAt": refresh_expires_at, "user": to_safe_user(user), "role": roles}

async def guest_register(db: AsyncSession, dto: dict) -> dict:
    existing = await user_provider.find_by_phone(db, dto["phone"])
    if existing:
        user = existing
    else:
        user = await user_provider.create_guest(db, name=dto["name"], phone=dto["phone"], is_guest=True)
    roles = ["GUEST"]
    tokens = await _issue_token_pair(db, user, roles, {})
    expires_at = (datetime.now(timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRY_MINUTES)).isoformat()
    return {"token": tokens["accessToken"], "expiresAt": expires_at, "user": to_safe_user(user), "role": roles}

async def login(db: AsyncSession, dto: dict, req_ctx: dict = {}) -> dict:
    user = await user_provider.find_by_email(db, dto["email"])
    if not user or not user.password:
        raise AppError(401, "Invalid email or password")
    if not verify_password(dto["password"], user.password):
        raise AppError(401, "Invalid email or password")
    roles = [ur.role.name for ur in user.roles if not ur.deleted] if user.roles else ["CUSTOMER"]
    if not roles:
        roles = ["CUSTOMER"]
    tokens = await _issue_token_pair(db, user, roles, req_ctx)
    expires_at = (datetime.now(timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRY_MINUTES)).isoformat()
    refresh_expires_at = (datetime.now(timezone.utc) + timedelta(days=REFRESH_TOKEN_EXPIRY_DAYS)).isoformat()
    return {"token": tokens["accessToken"], "expiresAt": expires_at, "refreshToken": tokens["refreshToken"], "refreshTokenExpiresAt": refresh_expires_at, "user": to_safe_user(user), "role": roles}

async def refresh_tokens(db: AsyncSession, raw_token: str, req_ctx: dict = {}) -> dict:
    hashed = hash_token(raw_token)
    stored = await refresh_token_provider.find_by_token(db, hashed)
    if not stored:
        raise AppError(401, "Invalid refresh token")
    if stored.revoked:
        await refresh_token_provider.revoke_family(db, stored.token_family)
        raise AppError(401, "Security alert: token reuse detected")
    if stored.expires_at < datetime.now(timezone.utc):
        raise AppError(401, "Refresh token expired")
    user = await user_provider.find_by_id(db, stored.user_id)
    if not user or user.deleted or not user.active:
        raise AppError(401, "User not found")
    await refresh_token_provider.revoke_by_id(db, stored.id, "rotation")
    roles = [ur.role.name for ur in user.roles if not ur.deleted] if user.roles else ["CUSTOMER"]
    tokens = await _issue_token_pair(db, user, roles, req_ctx)
    expires_at = datetime.now(timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRY_MINUTES)
    return {"accessToken": tokens["accessToken"], "expiresAt": expires_at.isoformat(), "expiresIn": ACCESS_TOKEN_EXPIRY_SECONDS, "refreshToken": tokens["refreshToken"], "refreshExpires": REFRESH_TOKEN_EXPIRY_SECONDS}

async def logout(db: AsyncSession, raw_token: str) -> None:
    if not raw_token:
        return
    hashed = hash_token(raw_token)
    stored = await refresh_token_provider.find_by_token(db, hashed)
    if stored:
        await refresh_token_provider.revoke_by_id(db, stored.id, "logout")

async def logout_all(db: AsyncSession, user_id: int) -> None:
    await db.execute(update(__import__('app.models.user', fromlist=['User']).User).where(__import__('app.models.user', fromlist=['User']).User.id == user_id).values(token_version=__import__('app.models.user', fromlist=['User']).User.token_version + 1))
    await db.commit()
    await refresh_token_provider.revoke_all_for_user(db, user_id)

async def change_password(db: AsyncSession, user_id: int, current_password: str, new_password: str, req_ctx: dict = {}) -> dict:
    from app.models.user import User as UserModel
    user = await user_provider.find_by_id(db, user_id)
    if not user:
        raise AppError(404, "User not found")
    if not verify_password(current_password, user.password or ""):
        raise AppError(401, "Current password is incorrect")
    if len(new_password) < 8:
        raise AppError(400, "New password must be at least 8 characters")
    new_hash = hash_password(new_password)
    await db.execute(update(UserModel).where(UserModel.id == user_id).values(password=new_hash, token_version=UserModel.token_version + 1))
    await db.commit()
    await refresh_token_provider.revoke_all_for_user(db, user_id)
    updated_user = await user_provider.find_by_id(db, user_id)
    roles = [ur.role.name for ur in updated_user.roles if not ur.deleted] if updated_user and updated_user.roles else ["CUSTOMER"]
    tokens = await _issue_token_pair(db, updated_user, roles, req_ctx)
    return tokens

async def me(db: AsyncSession, user_id: str) -> dict:
    user = await user_provider.find_by_id(db, int(user_id))
    if not user:
        raise AppError(404, "User not found")
    roles = [ur.role.name for ur in user.roles if not ur.deleted] if user.roles else ["CUSTOMER"]
    return {**to_safe_user(user), "role": roles or ["CUSTOMER"]}
